### ISSUE TITLE
Large files within source tree causes build to run out of memory

### ISSUE THREAD

**Issue by gulrotkake** [NixOS/nix#358]
If the size of a file or a sufficient number of files in a nix-build exceeds the amount of memory available, Nix runs out of memory and gives up. For example, given this simple build:

```
{ stdenv, pkgs, ... } :

stdenv.mkDerivation rec {
  name = "huge";
  src = ./.;
  buildInputs = [ ];
  buildPhase =
    ''
        md5sum huge.bin > md5.result
    '';

  installPhase =
   ''
     mkdir $out
     mv md5.result $out
  '';
}
```

where huge.bin is a very huge file, nix will first warn:

```
warning: dumping very large path (> 256 MiB); this may run out of memory
```

before eventually running out of memory.

The above was tested with

```
nix-build (Nix) 1.8pre3834_9b146a5
```

**Comment by domenkozar** [comment]
It would be nice if it printed out the file name

**Comment by sjmackenzie** [comment]
Agreed, the file name would be most handy.

**Comment by sjmackenzie** [comment]
We have 50 instances of `src = ./.` each time the `./.` points to the root of a rust project which consists of 4 files, one of which is in a folder called `src`. We have no instances of a large file/binary called with `./.`, just many (and growing quickly) instances of (`./.`).

**Comment by Ericson2314** [comment]
Why does it even read the files all into memory? I assume the hash can be computed with constant space.

**Comment by sjmackenzie** [comment]
For the record, we found the cause: it was a `.target` folder generated by the `rustc` compiler. Deleting that folder straightened `nix` again.

**Comment by domenkozar** [comment]
@Ericson2314 https://github.com/NixOS/nix/pull/619

**Comment by domenkozar** [comment]
Until this is fixed in Nix see https://nixos.org/wiki/How_to_add_files_to_the_nix-store#Large_files

**Comment by ip1981** [comment]
> As a workaround, you could use fetchurl to get the file, e.g. fetchurl { src = file:///bla/huge; sha256 = ...; }.

There might be permission problem.

**Comment by fisx** [comment]
> Until this is fixed in Nix see https://nixos.org/wiki/How_to_add_files_to_the_nix-store#Large_files

dead link, and it looks like this page does not exist on https://nixos.wiki/?

**Comment by dpc** [comment]
Is there any way to make `src = ./. except for ./target`?

**Comment by sjmackenzie** [comment]
take a look at `cleanSource` and `filterSource`

**Comment by siddharthist** [comment]
I am also experiencing this problem while trying to install Mathematica.

Here's the old Wiki page on archive.org in case anyone else is searching for it: https://web.archive.org/web/20160829175307/https://nixos.org/wiki/How_to_add_files_to_the_nix-store

**Comment by bgamari** [comment]
This is addressed by #619.

**Comment by Pitometsu** [comment]
Looks as a solution for me: https://stackoverflow.com/a/39493330/3195266
